src/com/ms/silverking/cloud/meta/ValueWatcher.java

```diff
@@ -7,87 +7,85 @@
-        this.listener = listener;
-        lastNotifiedZXID = Long.MIN_VALUE;
-    }
-    
-    public ValueWatcher(MetaClientCore metaClientCore, String basePath, ValueListener listener, 
-                          long intervalMillis) {
-        this(metaClientCore, basePath, listener, intervalMillis, intervalMillis);
-    }
-    
-    protected void _doCheck() throws KeeperException {
-        try {
-            ZooKeeperExtended   _zk;
-            byte[]              value;
-            Stat                stat;
-            
-            if (verbose) {
-                Log.warning("ValueCheck start: ", basePath);
-            }
-            _zk = metaClientCore.getZooKeeper();
-            stat = new Stat();
-            try {
-                value = _zk.getData(basePath, this, stat);
-            } catch (InterruptedException e) {
-                throw new RuntimeException(e);
-            }
-            
-            if (stat.getMzxid() > lastNotifiedZXID) {
-                listener.newValue(basePath, value, stat);
-                lastNotifiedZXID = stat.getMzxid();
-            }
-            if (verbose) {
-                Log.warning("ValueCheck complete: ", basePath);
-            }
-        } catch (KeeperException ke) {
-            System.out.println("*** ZooKeeper state: "+ metaClientCore.getZooKeeper().getState());
-            throw ke;
-        }
-    }
-    
-    private void checkValue() {
-        try {
-            doCheck();
-        } catch (KeeperException ke) {
-            throw new RuntimeException(ke);
-        }
-    }
-    
-    public void nodeDataChanged(WatchedEvent event) {
-        Log.fine("nodeDataChanged");
-        checkValue();
-    }    
+  private final ValueListener listener;
+  private long lastNotifiedZXID;
 
-    public void connected(WatchedEvent event) {
-        Log.fine("connected");
-        checkValue();
-    }
-    
-    public void nodeCreated(WatchedEvent event) {
-        Log.fine("nodeCreated");
-        checkValue();
+  private static final boolean verbose = false;
+
+  public ValueWatcher(MetaClientCore metaClientCore, String basePath, ValueListener listener, long intervalMillis,
+      long maxInitialSleep) {
+    super(metaClientCore, basePath, intervalMillis, maxInitialSleep);
+    this.listener = listener;
+    lastNotifiedZXID = Long.MIN_VALUE;
+  }
+
+  public ValueWatcher(MetaClientCore metaClientCore, String basePath, ValueListener listener, long intervalMillis) {
+    this(metaClientCore, basePath, listener, intervalMillis, intervalMillis);
+  }
+
+  protected void _doCheck() throws KeeperException {
+    try {
+      ZooKeeperExtended _zk;
+      byte[] value;
+      Stat stat;
+
+      if (verbose) {
+        Log.warning("ValueCheck start: ", basePath);
+      }
+      _zk = metaClientCore.getZooKeeper();
+      stat = new Stat();
+      try {
+        value = _zk.getData(basePath, this, stat);
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)